gguf-py/gguf/gguf_manager.py

```diff
@@ -0,0 +1,254 @@
+
+    # need to handle arrays separately
+    def add_array(self, key: str, val: Sequence[Any]) -> None:
+        if not isinstance(val, Sequence):
+            raise ValueError(f'Expected a sequence for {key}, got {type(val)}')
+        self.kv_data[key] = (val, GGUFValueType.ARRAY)
+
+    def add_tensor(
+        self, name: str, tensor: np.ndarray[Any, Any], raw_shape: Sequence[int] | None = None,
+        raw_dtype: GGMLQuantizationType | None = None,
+    ) -> None:
+        # we build splits as tensors are added so we need logic to figure out when to split
+        # logic is all in the conditional because it short-circuits, otherwise accessing self.shards[-1] would throw an error
+
+        # create a first shard to start it off
+        if (len(self.shards) == self.split_arguments.small_first_shard \
+            # or split when over tensor limit
+            or (self.split_arguments.split_style == SplitStyle.TENSORS \
+                and self.shards[-1][1] >= self.split_arguments.split_max_tensors) \
+            # or split when over size limit
+            or (self.split_arguments.split_style == SplitStyle.SIZE \
+                and self.shards[-1][2] + GGUFManager.get_tensor_size(tensor) > self.split_arguments.split_max_size)):
+
+            # we fill in the name later when we know how many shards there are
+            self.shards.append(["", 1, GGUFManager.get_tensor_size(tensor), deque([(name, tensor, raw_dtype)])])
+        else:
+            self.shards[-1][1] += 1
+            self.shards[-1][2] += GGUFManager.get_tensor_size(tensor)
+            self.shards[-1][3].append((name, tensor, raw_dtype))
+
+    def close(self) -> None:
+        for writer in self.shard_writers:
+            writer.close()
+
+    @staticmethod
+    def get_tensor_size(tensor) -> int:
+        try:
+            return tensor.data_type.elements_to_bytes(np.prod(tensor.shape))
+        except AttributeError: # numpy ndarray[Any, Any]
+            return tensor.nbytes
+        except: # this should never happen
+            raise ValueError(f"Invalid tensor type: {type(tensor)}")
+    
+    @staticmethod
+    def split_str_to_n_bytes(split_str: str) -> int:
+        if split_str.endswith("K"):
+            n = int(split_str[:-1]) * 1024
+        elif split_str.endswith("M"):
+            n = int(split_str[:-1]) * 1024 * 1024
+        elif split_str.endswith("G"):
+            n = int(split_str[:-1]) * 1024 * 1024 * 1024
+        elif split_str.isnumeric():
+            n = int(split_str)
+        else:
+            raise ValueError(f"Invalid split size: {split_str}, must be a number, optionally followed by K, M, or G")
+
+        if n <= 0:
+            raise ValueError(f"Invalid split size: {split_str}, must be positive")
+
+        return n
+
+    @staticmethod
+    def format_n_bytes_to_str(num: int) -> str:
+        if num == METADATA_ONLY_INDICATOR:
+            return "negligible - metadata only"
+        num = float(num)
```

Comment:
The `num` variable was declared to be an `int`, so a `float` should not be assigned to it. Use another variable for this.

Reply:
Resolved with 1cbab22.